File: auntieos-admin/android/app/src/main/java/com/tribetails/auntieos/domain/InvoiceReminder.kt
package com.tribetails.auntieos.domain

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * What the server decided when the admin asked for an invoice reminder (#832).
 *
 * `sendInvoiceReminder` refuses a second reminder inside its window and answers
 * `sent = false` with the time of the one that already went out. That is a
 * success: the household has been reminded. The app says so, and when, instead
 * of "Reminder sent." for a send that did not happen.
 */
data class ReminderOutcome(
    /** True when THIS call sent a reminder. */
    val sent: Boolean,
    /** When the most recent reminder went out (ms epoch). */
    val lastReminderAtMs: Long,
    /** The earliest moment another reminder will be accepted (ms epoch). */
    val nextReminderAllowedAtMs: Long,
)

/** "Sep 14, 3:05 PM", in [zone] (the device's own zone unless a test pins one). */
fun formatReminderTime(ms: Long, zone: TimeZone = TimeZone.getDefault()): String =
    SimpleDateFormat("MMM d, h:mm a", Locale.US).apply { timeZone = zone }.format(Date(ms))

/** The toast or banner line after a press. */
fun reminderOutcomeMessage(outcome: ReminderOutcome, zone: TimeZone = TimeZone.getDefault()): String =
    if (outcome.sent) {
        "Reminder sent."
    } else {
        "Not sent: a reminder already went out ${formatReminderTime(outcome.lastReminderAtMs, zone)}. " +
            "The next one can go out after ${formatReminderTime(outcome.nextReminderAllowedAtMs, zone)}."
    }

/** The value of the "Last reminder" row. Absent, null or non-positive means none on record. */
fun lastReminderLabel(stampMs: Long?, zone: TimeZone = TimeZone.getDefault()): String =
    if (stampMs != null && stampMs > 0L) formatReminderTime(stampMs, zone) else "none sent"

File: auntieos-admin/src/lib/invoiceReminder.ts
/**
 * What happened when the admin asked for an invoice reminder, and how to say it
 * (#832).
 *
 * `sendInvoiceReminder` refuses a second reminder inside its window and answers
 * `sent: false` with the time of the one that already went out. That answer is a
 * success, not an error: the household has been reminded. The panel says so in a
 * sentence and shows when, instead of claiming "Reminder sent." for a send that
 * did not happen.
 */
import type { SendInvoiceReminderResult } from '../contracts/invoiceContracts.generated';

export interface ReminderOutcome {
  /** True when THIS call sent a reminder. */
  sent: boolean;
  /** When the most recent reminder went out (ms epoch). */
  lastReminderAtMs: number;
  /** The earliest moment another reminder will be accepted (ms epoch). */
  nextReminderAllowedAtMs: number;
}

/**
 * Decodes the callable's answer.
 *
 * A response WITHOUT `sent` comes from a function deployed before #832, which
 * only ever answered after sending. Reading that as "not sent" would tell the
 * admin a reminder was refused when it went out, so it reads as sent, now.
 */
export function reminderOutcomeOf(
  res: Partial<SendInvoiceReminderResult> | null | undefined,
  nowMs: number,
): ReminderOutcome {
  if (typeof res?.sent !== 'boolean') {
    return { sent: true, lastReminderAtMs: nowMs, nextReminderAllowedAtMs: nowMs };
  }
  const last = typeof res.lastReminderAtMs === 'number' ? res.lastReminderAtMs : nowMs;
  const next = typeof res.nextReminderAllowedAtMs === 'number' ? res.nextReminderAllowedAtMs : last;
  return { sent: res.sent, lastReminderAtMs: last, nextReminderAllowedAtMs: next };
}

/** "Sep 14, 3:05 PM". `timeZone` is for tests; the panel uses the viewer's zone. */
export function formatReminderTime(ms: number, timeZone?: string): string {
  return new Date(ms).toLocaleString('en-US', {
    month: 'short',
    day: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
    ...(timeZone ? { timeZone } : {}),
  });
}

/** The notice after a press. */
export function reminderOutcomeMessage(outcome: ReminderOutcome, timeZone?: string): string {
  if (outcome.sent) return 'Reminder sent.';
  return `Not sent: a reminder already went out ${formatReminderTime(outcome.lastReminderAtMs, timeZone)}. The next one can go out after ${formatReminderTime(outcome.nextReminderAllowedAtMs, timeZone)}.`;
}

/**
 * The value of the "Last reminder" fact. The stored stamp may be absent, null
 * (a released claim), or not a number on a doc written outside the callables;
 * all of those mean no reminder is on record.
 */
export function lastReminderLabel(stamp: unknown, timeZone?: string): string {
  return typeof stamp === 'number' && Number.isFinite(stamp) && stamp > 0
    ? formatReminderTime(stamp, timeZone)
    : 'none sent';
}
